### PR TITLE
Generic subscriptions

### PR DESCRIPTION
**Pull request by DanMesh**
Allows adding subscriptions to arbitrary ROS topics.
Users can register (multiple) callbacks or simply poll the last-received message.

### Example usage

```cpp
px4_ros2::Subscription<px4_msgs::msg::VehicleLocalPosition> ros_sub(context, "/fmu/out/vehicle_local_position");

// Create callback
ros_sub.onUpdate([](px4_msgs::msg::VehicleLocalPosition msg) {
    std::cout << "Local pos from callback: " << msg.x << " " << msg.y << " " << msg.z << std::endl;
});

// Poll latest message
px4_msgs::msg::VehicleLocalPosition msg = ros_sub.last();
std::cout << "Local pos latest: " << msg.x << " " << msg.y << " " << msg.z << std::endl;
```

### CODE REVIEW

File: px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        }
+      });
+  }
+
+  /**
+   * @brief Register a callback to be executed for each message.
+   *
+   * @param callback the callback function
+   */
+  void subscribe(std::function<void(RosMessageType)> callback)
```
Comment (DanMesh) [inline px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp:42]: Perhaps a better name is `onUpdate()` or `addCallback()`?

File: px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  {
+    _callbacks.push_back(callback);
+  }
+
+  /**
+   * @brief Get the last-received message.
+   *
+   * @returns the last-received ROS message
+   * @throws std::runtime_error when no messages have been received
```
[the remark this answers is not in the document]
Comment (DanMesh) [reply px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp:51]: How shall we define "valid"?
1. A message has been received at _any_ point in the past, or
2. a message was received within a certain time (e.g. 500ms).

[the remark this answers is not in the document]
Comment (bkueng) [reply px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp:51]: Being a bit specific here probably isn't a bad idea, so I'd go for 2 with a default duration settable via argument (something around 500ms-1s default).

[the remark this answers is not in the document]
Comment (DanMesh) [reply px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp:54]: I added `lastValid()` with a default of 500ms

File: px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp (hunk 1 above)
Comment (DanMesh) [reply px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp]: Renamed to `onUpdate`.

I made the global & local position classes inherit from `Subscription` so they'll have this and other functions.

File: px4_ros2_cpp/include/px4_ros2/odometry/local_position.hpp [hunk 3]
```diff
@@ -20,48 +19,47 @@ namespace px4_ros2
...
+class OdometryLocalPosition : public Subscription<px4_msgs::msg::VehicleLocalPosition>
 {
 public:
   explicit OdometryLocalPosition(Context & context);
 
-  bool positionXyValid() const
+  bool positionXYValid() const
   {
-    return _vehicle_local_position.xy_valid;
+    return last().xy_valid;
```
Comment (bkueng) [inline px4_ros2_cpp/include/px4_ros2/odometry/local_position.hpp:29]: (below too)
```suggestion
    return lastValid() && last().xy_valid;
```

File: px4_ros2_cpp/include/px4_ros2/odometry/local_position.hpp [hunk 4]
```diff
@@ -20,48 +19,47 @@ namespace px4_ros2
...
   bool positionZValid() const
   {
-    return _vehicle_local_position.z_valid;
+    return last().z_valid;
   }
 
   Eigen::Vector3f position() const
   {
-    return {_vehicle_local_position.x, _vehicle_local_position.y, _vehicle_local_position.z};
+    const px4_msgs::msg::VehicleLocalPosition pos = last();
```
Comment (bkueng) [inline px4_ros2_cpp/include/px4_ros2/odometry/local_position.hpp:39]:
```suggestion
    const px4_msgs::msg::VehicleLocalPosition& pos = last();
```

File: px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        }
+      });
+  }
+
+  /**
+   * @brief Add a callback to execute when receiving a new message.
+   *
+   * @param callback the callback function
+   */
+  void onUpdate(std::function<void(RosMessageType)> callback)
```
Comment (bkueng) [inline px4_ros2_cpp/include/px4_ros2/odometry/subscription.hpp:45]: You could add a `using UpdateCallback = std::function<void(const RosMessageType&)>`

```suggestion
  void onUpdate(const std::function<void(const RosMessageType&)>& callback)
```

File: px4_ros2_cpp/include/px4_ros2/odometry/global_position.hpp [hunk 6]
```diff
@@ -22,27 +19,16 @@ namespace px4_ros2
...
 {
 public:
   explicit OdometryGlobalPosition(Context & context);
 
-  bool valid()
-  {
-    return _node.get_clock()->now() - _last_vehicle_global_position < 500ms;
-  }
-
   Eigen::Vector3d position() const
```
[the remark this answers is not in the document]
Comment (DanMesh) [reply px4_ros2_cpp/include/px4_ros2/odometry/global_position.hpp:27]: I'll add this for local positions, but for the global position, NED isn't quite right.

Should we add something to specify it's lat-lon-alt?

[the remark this answers is not in the document]
Comment (bkueng) [reply px4_ros2_cpp/include/px4_ros2/odometry/global_position.hpp:27]: Ah yes, I was thinking this was the local pos.
Adding a comment should be enough then here.

**Review by bkueng — APPROVED** [review]
Nice

**Pull request merged**